Allow sync only on string/hashtag

gregharvey opened the pull request:
Fixes #7 - I've tested this with the `sync_hashtag` set to empty (everything syncs as usual) and I've tried it both ways with a hashtag. You can actually set different hashtags for Twitter and Mastodon if you want, and it will work with a string too - it doesn't *have* to be a hashtag either, I just left it like that in the docs because hashtags are what people are used to.

gregharvey commented:
Oh, another advantage of separate config for Twitter and Mastodon is you can make everything sync one way and have selective syncing the other way, e.g. everything I tweet goes to Mastodon but only things I tag on Mastodon go to Twitter.

gregharvey commented:
That's weird. It compiles locally! Must be different rust versions?

gregharvey commented:
Lol, compiler is clearly trolling me. It compiles locally both ways. I add `"".to_string()` and it fails saying I can't use that method in that context (even though it compiles locally). I remove it and it fails saying `help: try using a conversion method: "".to_string()` ...

Suggestions, anybody? I'm very new to rust!

gregharvey commented:
Well, I don't know what to say. I just compiled locally with this:

`cargo build --verbose --release`

Compiled locally with no issues:

```
   Compiling mastodon-twitter-sync v1.4.0 (/home/greg/workspace/github/mastodon-twitter-sync)
     Running `rustc --crate-name mastodon_twitter_sync --edition=2018 src/main.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type bin --emit=dep-info,link -C opt-level=3 -C lto -C metadata=221888c4170a3397 -C extra-filename=-221888c4170a3397 --out-dir /home/greg/workspace/github/mastodon-twitter-sync/target/release/deps -L dependency=/home/greg/workspace/github/mastodon-twitter-sync/target/release/deps --extern chrono=/home/greg/workspace/github/mastodon-twitter-sync/target/release/deps/libchrono-520344bded9cecf7.rlib --extern dissolve=/home/greg/workspace/github/mastodon-twitter-sync/target/release/deps/libdissolve-8294b5f69fe8fc1e.rlib --extern egg_mode=/home/greg/workspace/github/mastodon-twitter-sync/target/release/deps/libegg_mode-85077033a35858fd.rlib --extern egg_mode_text=/home/greg/workspace/github/mastodon-twitter-sync/target/release/deps/libegg_mode_text-6b198d2e8319927b.rlib --extern env_logger=/home/greg/workspace/github/mastodon-twitter-sync/target/release/deps/libenv_logger-87a64b833e783b1d.rlib --extern failure=/home/greg/workspace/github/mastodon-twitter-sync/target/release/deps/libfailure-e2d700f2157ad9ba.rlib --extern log=/home/greg/workspace/github/mastodon-twitter-sync/target/release/deps/liblog-15deecde67a2dad9.rlib --extern mammut=/home/greg/workspace/github/mastodon-twitter-sync/target/release/deps/libmammut-410d63e492a96011.rlib --extern mime=/home/greg/workspace/github/mastodon-twitter-sync/target/release/deps/libmime-aebb44b8e0ed9959.rlib --extern regex=/home/greg/workspace/github/mastodon-twitter-sync/target/release/deps/libregex-e372e672412c6452.rlib --extern reqwest=/home/greg/workspace/github/mastodon-twitter-sync/target/release/deps/libreqwest-8f1150e6a734bb70.rlib --extern serde=/home/greg/workspace/github/mastodon-twitter-sync/target/release/deps/libserde-d2a5cd2f38d401dc.rlib --extern serde_json=/home/greg/workspace/github/mastodon-twitter-sync/target/release/deps/libserde_json-cdfab80522e60686.rlib --extern structopt=/home/greg/workspace/github/mastodon-twitter-sync/target/release/deps/libstructopt-9b5195f73a7472d6.rlib --extern tempfile=/home/greg/workspace/github/mastodon-twitter-sync/target/release/deps/libtempfile-036c9b1376750828.rlib --extern tokio=/home/greg/workspace/github/mastodon-twitter-sync/target/release/deps/libtokio-ae7cb6f50fca480f.rlib --extern toml=/home/greg/workspace/github/mastodon-twitter-sync/target/release/deps/libtoml-a9ba883add2b5c94.rlib -L native=/home/greg/workspace/github/mastodon-twitter-sync/target/release/build/backtrace-sys-130352b344b469a3/out`
    Finished release [optimized] target(s) in 12m 16s
```

So I really don't know what to do to make Travis happy!

klausi commented:
You need to run `cargo test` which will execute the automated tests, then it fails to compile. Looks like there is a problem in the test code which is not compiled for the release binary.

gregharvey commented:
Ahh, OK - thanks. I'll have to get my head around how the tests work, it's my first foray into rust! :-)

Bear with me, I'll get these tests added...

gregharvey commented:
Updated the README and switched to using the Option type for the hashtag variables instead of String. This seems to be working for me, and it compiles now too. I am yet to write the tests themselves, will do that next.

gregharvey commented:
Added three tests:

- tagged posts are sent when config is in place
- untagged posts are _not_ sent when config is in place
- all posts are sent when hashtag config is not present (Option is None)

gregharvey commented:
Thinking about it, the third test is unnecessary, as *all the other tests* necessarily test that case. Oops! Let me know and I'll remove it. Unless you think it's worthwhile anyway?

klausi commented on src/registration.rs:
```diff
@@ -60,6 +60,7 @@ pub async fn twitter_register() -> Result<TwitterConfig> {
             delete_older_statuses: false,
             delete_older_favs: false,
             sync_retweets: true,
+            sync_hashtag: std::option::Option::Some("".to_string()),
```
This should be set to Option::None per default.

klausi commented on src/sync.rs:
```diff
@@ -56,11 +58,24 @@ pub fn determine_posts(
+        // Fetch the tweet text into a String object
+        let decoded_tweet = tweet_unshorten_decode(tweet);
+        // Fetch the sync hashtag from the Option object for comfort
+        let sync_hashtag_twitter_value = options.sync_hashtag_twitter.as_deref().unwrap_or("");
```
we should change this to `match options.sync_hashtag_twitter` and then have 2 arms for the None and for the Some case.

klausi commented on src/sync.rs:
```diff
@@ -85,11 +100,22 @@ pub fn determine_posts(
+        // The toot is not on Twitter yet, next step
+
+        // Fetch the sync hashtag from the Option object for comfort
+        let sync_hashtag_mastodon_value = options.sync_hashtag_mastodon.as_deref().unwrap_or("");
```
same here, match.

klausi commented on src/sync.rs:
```diff
@@ -414,6 +440,8 @@ mod tests {
     static DEFAULT_SYNC_OPTIONS: SyncOptions = SyncOptions {
         sync_reblogs: true,
         sync_retweets: true,
+        sync_hashtag_twitter: std::option::Option::None,
```
can just be `None` without the fully qualified name, as we use otherwise in the code base. Please do that for all Option/Some/None.

klausi commented on src/sync.rs:
```diff
@@ -936,6 +964,62 @@ QT test123: Original text"
+        assert!(posts.tweets.is_empty());
+    }
+
+    // Test all posts are sent if hashtag config is unset
```
Agreed, this test is already covered, so can be removed.

klausi commented:
Ah, I'm a bit worried about the config. Can we make a test in config.rs that the empty string for the hashtag ends up as `None` in our datastructure?

klausi commented:
Merged your changes and refactored it according to my comments. Thank you!
